### DISCUSSION TITLE
Is the LEGO IR USB Tower Supported?

### DISCUSSION THREAD

**Discussion by bamartin125** [maehw/WebPBrick#8] [Q&A] [answered]
Is the LEGO IR USB Tower Supported as of now? I am attempting to connect to the tower via webprick.com using a Chrome browser on a Manjaro Linux distribution (Linux 6.1.100).

I did happen to find [this post](https://www.eurobricks.com/forum/index.php?/forums/topic/197807-webpbrickcom-programming-the-1st-generation-lego-mindstorms-ris-yellow-rcx-brick-from-the-webbrowser/&do=findComment&comment=3694162) made back in April indicating there were problems in the past. However, since this discussion forum exists, I figured I would verify that there are still issues with this connection and get the current state documented here.

My apologies if this question has been posted in the wrong area.

**Comment by maehw** [comment, marked as answer]
Hi @bamartin125,

thanks for your interest in WebPBrick and its support for the LEGO IR USB Tower.

TL;DR: Unfortunately, as of today, the USB tower is not fully supported and only in an experimental state. I guess I need support from somebody who's interested+motivated in WebUSB + JavaScript async in getting this to run. I am even willing to provide equipment (USB tower + RCX programmable brick).

---

Since there was zero feedback from the eurobricks community, I haven't kept my post up to date and only updated the [German forum post on 1000steine.de](https://www.1000steine.de/de/gemeinschaft/forum/?entry=1&id=477706#id477706).

You'd need to clone the git repository's main branch and navigate to: ./src/communication/webusb.htm - this page provides a PoC. IIRC it should even try to download firmware on the RCX (so use at your own risk if you currently have running firmware (and potentially also user programs) running on the brick). We seem to loose data somewhere between the RCX and the webbrowser - I don't know if the data is already lost on the USB or not.

HTH, feedback is always welcome.

**Reply by n-davey** [reply]
I stumbled across this when looking for options to use the RCX brick.. Looks nice, going to give it a try... 
I'm might look at the USB tower interface too.. for some reason this has piqued my interest..

**Reply by maehw** [reply]
Hi @n-davey ,

that sounds good. Do you have both an IR serial and an IR USB tower at hand?

If you had a Saleae Logic analyzer at hand, my new tool might come in handy in analyzing the USB traffic between tower and host PC: https://github.com/maehw/UsbInsight

Otherwise, different OSes provide different means of sniffing USB traffic. It could help to see what's the difference is when the USB tower is controlled by Bricx Command Center (BricxCC) and by the WebUSB code. 

I am looking forward to hear from you. 

Have a nice weekend

**Reply by n-davey** [reply]
Hi [maehw](https://github.com/maehw),

I have a working serial tower and 2 untested usb towers on hand yes.
I do also own a Saleae Logic Pro 16..

Now to find a little time to play..
